### ISSUE TITLE
[e2e] Document the best path for making changes to simple-game-server for e2e tests

### ISSUE THREAD

**Issue by markmandel** [googleforgames/agones#3836]
**Is your feature request related to a problem? Please describe.**

There is lots of confusion on the best path for how to make changes to simple game server in a safe way.

**Describe the solution you'd like**

Make the ImagePullPolicy a variable, so it can easily be set to `Always`:

Variables:

https://github.com/googleforgames/agones/blob/bff72b1161492d7386e0bf39ca0753802c4b3e3a/test/e2e/framework/framework.go#L152-L169

Line:

https://github.com/googleforgames/agones/blob/bff72b1161492d7386e0bf39ca0753802c4b3e3a/test/e2e/framework/framework.go#L802

Write a new section in https://github.com/googleforgames/agones/blob/main/build/README.md#testing-and-building called "Making changes to Simple Game Server with e2e Tests", with the following documentation:

1. How to create your own local image for Simple Game Server, and point to https://github.com/googleforgames/agones/blob/main/build/README.md#running-individual-end-to-end-tests for beginning of making changes.
2. For moving to CI - set the [image tag](https://github.com/googleforgames/agones/blob/bff72b1161492d7386e0bf39ca0753802c4b3e3a/examples/simple-game-server/Makefile#L41) to a `0.xx-dev` tag (so `-dev` with the version number), and `imagePullPolicy` to `Always`, and create a **draft** PR. 
3. With this approach the developer (or maintainer who is helping) can publish the `-dev` version multiple times without polluting the production version space, can can fix issues with CI's help without any risk.
4. Once CI is passing and review has occured, the PR can be edited to remove the `-dev` tag, move `imagePullPolicy` back to `IfNotPresent`, and the production version `0.xx` can be published -- and the PR merged.
5. The `-dev` can also be manually cleaned up if required as well.

**Describe alternatives you've considered**

Rather than a `-dev` suffix, we could guide people to use the `us-docker.pkg.dev/agones-images/ci` repository, that will have a 40 day cleanup rule (eventually - I need to check the audit logs for some previous cleanup work), so no manual cleanup step is required.

So, for example - `us-docker.pkg.dev/agones-images/examples/simple-game-server:0.32` would become `us-docker.pkg.dev/agones-images/ci/simple-game-server:0.32`

I think I like this better - but would like consensus.

**Additional context**

N/A

**Comment by markmandel** [comment]
@ashutosji - since you are working on this, would love your feedback.

**Comment by markmandel** [comment]
Moving to `awaiting-maintainer` since we should really do this.

**Comment by kamaljeeti** [comment]
> **Describe alternatives you've considered**
> 
> Rather than a `-dev` suffix, we could guide people to use the `us-docker.pkg.dev/agones-images/ci` repository, that will have a 40 day cleanup rule (eventually - I need to check the audit logs for some previous cleanup work), so no manual cleanup step is required.

@markmandel I’m leaning toward the second approach using the CI-specific registry (agones-images/ci)—it sounds more maintainable since it avoids manual cleanup by relying on the 40-day cleanup rule. However, I’m not entirely sure how that cleanup policy works in practice. 
Could you please provide more details or point me to the audit logs from previous cleanups? Just want to understand how reliable and automated it actually is.

Curious to hear what others think—is the CI approach stable enough to adopt as a standard, or should we stick to the -dev tagging strategy for now?

**Comment by markmandel** [comment]
https://cloud.google.com/artifact-registry/docs/repositories/cleanup-policy#delete

I don't have access anymore to the Cloud project, but you can go look.

**Comment by Sivasankaran25** [comment]
Hi @markmandel 

Here’s the solution description I’m proposing — can you confirm if this correctly reflects what we want to implement?

### ***Solution Description***

We will add a configurable _image-pull-policy_ flag to the Agones e2e testing framework, so developers can easily choose _IfNotPresent, Always, or Never_ without code changes.

We’ll also add a new "Making Changes to Simple Game Server with e2e Tests" section in **build/README.md** that documents:

1. Building and publishing a development Simple Game Server image
2. Running e2e tests with _image-pull-policy=Always_ to ensure fresh pulls.
3. Two options for temporary image storage:
                  - Preferred: CI-specific registry _(us-docker.pkg.dev/agones-images/ci/...)_ with 40-day auto-cleanup.
                  - Alternative: -dev suffix tagging for manual cleanup.
4. Safe transition from dev to production images without polluting stable tags.

This will make it easier and safer to test game server changes, iterate in CI, and avoid production tag pollution.

**Comment by markmandel** [comment]
> Running e2e tests with image-pull-policy=Always to ensure fresh pulls.

🤔 do we need this, or should we have just new `-dev` suffix or `+meta` information on the tag for each build? (Especially if it's cleaning up after 40 days)?

**Comment by indurireddy-TF** [comment]
Hi @markmandel Requiring developers to manually create a new tag for every quick fix creates high friction and a risk of testing stale, cached code. Setting ImagePullPolicy=Always once simplifies the workflow, reliably guaranteeing that the latest code is pulled and tested every time.

**Comment by markmandel** [comment]
@indurireddy-TF - should we just do #4329 instead? Then we'd have a tag per commit and could clean them up with automated rules on the artifact registry.

Seems like the easiest option?

**Comment by indurireddy-TF** [comment]
@markmandel I completely agree. That automated solution is much better and removes all the manual hassle we are facing.

**Comment by markmandel** [comment]
To be honest - no idea why the idea didn't occur in the past 😄

**Comment by markmandel** [comment]
Any objection to closing this issue as duplicate of #4329 ?

**Comment by indurireddy-TF** [comment]
Yes, we can close this.
